### PR TITLE
Email and SmsPreferred support in Altinn.Notifications.Core

### PR DESCRIPTION
**Pull request by acn-sbuad**
## Related Issue(s)
- #541

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.Notifications.Core/Services/Interfaces/IEmailOrderProcessingService.cs [hunk 1]
```diff
@@ -13,7 +14,19 @@ public interface IEmailOrderProcessingService
...
+    /// </summary>
+    public Task ProcessOrderWithoutAddressLookup(NotificationOrder order, List<Recipient> recipients);
+
+    /// <summary>
+    /// Retry processing of a notification order
     /// </summary>
     public Task ProcessOrderRetry(NotificationOrder order);
+
+    /// <summary>
+    /// Pretry processing of a notification order for the provided list of recipients
```
Comment (HenningNormann) [inline src/Altinn.Notifications.Core/Services/Interfaces/IEmailOrderProcessingService.cs:28]: Typoe?

File: src/Altinn.Notifications.Core/Services/Interfaces/ISmsOrderProcessingService.cs [hunk 2]
```diff
@@ -12,8 +13,20 @@ public interface ISmsOrderProcessingService
...
+    /// </summary>
+    public Task ProcessOrderWithoutAddressLookup(NotificationOrder order, List<Recipient> recipients);
+
     /// <summary>
     /// Retry processing of an order
     /// </summary>
     public Task ProcessOrderRetry(NotificationOrder order);
+
+    /// <summary>
+    /// Pretry processing of a notification order for the provided list of recipients
```
Comment (HenningNormann) [inline src/Altinn.Notifications.Core/Services/Interfaces/ISmsOrderProcessingService.cs:28]: Typoe?

File: src/Altinn.Notifications.Core/Services/PreferredChannelProcessingService.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+/// Implementation of the <see cref="IPreferredChannelProcessingService"/>
+/// </summary>
+public class PreferredChannelProcessingService : IPreferredChannelProcessingService
+{
+    private readonly IEmailOrderProcessingService _emailProcessingService;
+    private readonly ISmsOrderProcessingService _smsProcessingService;
+    private readonly IContactPointService _contactPointService;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="OrderProcessingService"/> class.
```
Comment (HenningNormann) [inline src/Altinn.Notifications.Core/Services/PreferredChannelProcessingService.cs:18]: copy/paste feil ved cref?

File: src/Altinn.Notifications.Core/Services/OrderRequestService.cs [hunk 4]
```diff
@@ -131,6 +118,46 @@ public async Task<NotificationOrderRequestResponse> RegisterNotificationOrder(No
...
+        };
+    }
+
+    private static List<Recipient> GetMissingContactRecipientList(NotificationChannel channel, List<Recipient> recipients)
+    {
+        return channel switch
+        {
+            NotificationChannel.Email => recipients
+                               .Where(r => !r.AddressInfo.Exists(ap => ap.AddressType == AddressType.Email))
+                               .Select(r => r.DeepCopy())
```
Comment (HenningNormann) [inline src/Altinn.Notifications.Core/Services/OrderRequestService.cs:147]: Why is DeepCopy necessary? Will the original input ever be used assumed unchanged?

Comment (acn-sbuad) [reply src/Altinn.Notifications.Core/Services/OrderRequestService.cs:147]: If we don't deep copy the recipient info we find during the lookup would be persisted with the notification order. We do not want to persists this information until the send time. We first perform the lookup (triggered from line 50) and then we use the same object to generate the notification order to be persisted in the db (line 67)

**Review by HenningNormann — APPROVED** [review]

**Pull request merged**